### PR TITLE
Add product search endpoint

### PR DESCRIPTION
**Pull request by BirgitBader**

### CODE REVIEW

File: _raml/apps/search/examples/post_search_products_request.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+    "query": "Boom",
+    "sortBy": "RELEVANCE"
+  }
```
Comment (mhoehn) [inline _raml/apps/search/examples/post_search_products_request.json]: Resource ist nur für Now:

{
  "query" : "Handbag",
  "sortBy" : "RELEVANCE",
  "attributeFilters" : [
    {
      "attributeValue" : null,
      "attributeRange" : {
        "min" : 1,
        "max" : 500
        
      },
      "fieldName" : "ListPrice"
    },
    {
      "attributeValue" : 22076,
      "fieldName" : "CategoryID"
    },
    {
      "attributeValue": "red",
      "fieldName": "PreDefString_de_67648890"
    }
  ]
}

File: _raml/apps/search/examples/post_search_products_response.json [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+          },
+          {
+            "rel": "categories",
+            "href": "https://pm.epages.com/rs/shops/apidocu/categories/?productId=51E7F913-5447-240E-77AE-AC14145FA477"
+          }
+        ]
+      }
+    ]
+  }
+
```
Comment (mhoehn) [inline _raml/apps/search/examples/post_search_products_response.json]: geändert: siehe /product/src/main/resources/raml_console/public/raml/facetedsearch/examples/search_output.json

File: _raml/apps/search/schema/post_search_products.schema.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    "$schema": "http://json-schema.org/draft-03/schema",
+    "title": "Post Search Product Schema",
+    "type": "object",
+    "properties": {
+      "": {
+        "type": "object of [product](page:apps-data-types#product)",
+        "description": ""
+      }
+    }
+  }
```
Comment (mhoehn) [inline _raml/apps/search/schema/post_search_products.schema.json]: siehe auch: /product/src/main/resources/raml_console/public/raml/facetedsearch/schema/search_input_schema.json

File: apps/data-types.md [hunk 4]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
 | lineItemPrice | object of [price](page:apps-data-types#price) | The price of the line item. |
 | lineItemProductGuid | string | The identifier of the related product line item. A deposit is always bound to a product.|
 
+## facet
+
+| Attribute | Type | Description |
+| - | - |  - |
+| attributeName | string | The name of the filter. Can be *ListPrice*, *Manufacturer*, *CategoryID*, or any other attribute created by the merchant.  |
+| attributeType | string | xxx |
```
Comment (mhoehn) [inline apps/data-types.md]: "Standard" for Manufacturer, Category or ListPrice
"PreDefString" for own attributes

File: apps/data-types.md [hunk 5]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
...
+| - | - |  - |
+| attributeValue  | string | The value of the facet. |
+| attributeRange  | object of [attributeRange](page:apps-data-types#attributerange) | The minimum and maximum range of the filter applied. |
+| fieldName  | string | The alias of the filter.  |
+
+## filterValues
+
+| Attribute | Type | Description |
+| - | - |  - |
+| matches | string | xxx |
```
Comment (mhoehn) [inline apps/data-types.md]: count of objects matched by this filter criteria

File: apps/data-types.md [hunk 6]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
...
+| name | string | xxx |
```
Comment (mhoehn) [inline apps/data-types.md]: localized name of attribute

File: apps/data-types.md [hunk 7]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
...
+| filter | object of [filter](page:apps-data-types#filter) | xxx |
+| selectedValue | string | xxx |
```
Comment (mhoehn) [inline apps/data-types.md]: This is the selected attribute value used as filter. (This value was given used in the input as filter value)

File: apps/data-types.md [hunk 8]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
...
+| attributeRange  | object of [attributeRange](page:apps-data-types#attributerange) | The minimum and maximum range of the filter applied. |
+| fieldName  | string | The alias of the filter.  |
+
+## filterValues
+
+| Attribute | Type | Description |
+| - | - |  - |
+| matches | string | xxx |
+| name | string | xxx |
+| filter | object of [filter](page:apps-data-types#filter) | xxx |
```
Comment (mhoehn) [inline apps/data-types.md]: available filter with name and value (or range)  usable as input parameter

File: _raml/apps/search/examples/post_search_products_response.json [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        "productVariationMasterId": null,
+        "manufacturerProductNumber": "CU001",
+        "productLength": null,
+        "productWidth": null,
+        "productHeight": null,
+        "productVariationSelection": null,
+        "shortDescription": null,
+        "deliveryPeriod": "2-4",
+        "description": "Customise your headphone in no time!",
+        "title": "Boom - Customized headphone - Technic365",
```
Comment (cgebken) [inline _raml/apps/search/examples/post_search_products_response.json]: Is it on purpose that you switch between en-GB and en-US spelling of customize in this PR? You also have a product called "Boom - Customised headphone" and the description of this product also uses "customise".

File: apps/data-types.md [hunk 10]
```diff
@@ -251,6 +259,31 @@ This object is used for the attributes of shippingAddress and billingAddress.
 | lineItemPrice | object of [price](page:apps-data-types#price) | The price of the line item. |
 | lineItemProductGuid | string | The identifier of the related product line item. A deposit is always bound to a product.|
 
+## facet
+
+| Attribute | Type | Description |
+| - | - |  - |
+| attributeName | string | The name of the filter. Can be *ListPrice*, *Manufacturer*, *CategoryID*, or any other attribute created by the merchant.  |
+| attributeType | string | The type of the filter. Is *Standard* for *Manufacturer*, *CategoryID*, or *Price*. Is *PreDefString* for any other attribute created by the merchant. |
```
Comment (cgebken) [inline apps/data-types.md]: In the definition of attributeName you mention ListPrice and here only Price. Isn't it the same?

File: apps/data-types.md [hunk 11]
```diff
@@ -540,6 +573,83 @@ This object is used for the attributes of basePrice, depositPrice, ecoParticipat
...
+| deliveryWeight | object of [deliveryWeightQuantity](page:apps-data-types#deliveryweightquantity) | The delivery weight of the product. |
+| shippingMethodsRestrictedTo | array of [link](page:apps-data-types#link) | Information on possible shipping method restrictions, e.g. express delivery only. Can be `null` if no restrictions exist. |
+| availabilityText | string | Additional custom information on the product's stock level or the delivery period. |
+| availability | string | The availability of the product. Can be one of *OnStock*, *WarnStock*, *OutStock*. |
+| energyLabelsMaxEfficiencyString | string | The highest possible efficiency class of an uploaded energy label. |
+| energyLabelsString | string | A list of energy labels applied to this product. Can be one value or a range with two values. |
+| energyLabelSourceFile | string | A URL with an image or PDF file containing the energy label image supplied by the manufacturer. |
+| productDataSheet | string | An image or PDF file containing a data sheet with technical information on the product. Has to be available if the product has an energy label. |
+| sfUrl | string | The link to storefront URL of the product. |
```
Comment (cgebken) [inline apps/data-types.md]: to THE storefront URL

File: apps/data-types.md [hunk 12]
```diff
@@ -540,6 +573,83 @@ This object is used for the attributes of basePrice, depositPrice, ecoParticipat
...
+| stocklevel | number | Indicates the stock level of the product. |
+| minStocklevel | number | Indicates the minimum stock level of the product. Available if the field *Minimum stock level* is used in the administration area of a shop. Only available with the `products_write` authorisation. |
+| links | array of [link](page:apps-data-types#link) | The links to the product and product category. |
+| productVariationType | string | The type of the variation product. Can be one of *master*, *regular* or *variation*. |
+| productVariationMasterId | string | The unique identifier of the variation product. Null if the `productVariationType` is *master* or *regular*. |
+| productVariationMasterName | string | The name of the variation product. Null if the `productVariationType` is *master* or *regular*. |
+| productVariationSelection | array of [variationIdentifier](page:apps-data-types#variationidentifier) | The selection of the variation product. Only available if `productVariationType` is *variation*. |
+| videos | array of [video](page:apps-data-types#video) | The videos belonging to the product. |
+| query | string | The search query. |
+| fallbackQuery | string | xxx |
```
Comment (cgebken) [inline apps/data-types.md]: Is xxx really the final description?

File: _raml/apps/search/examples/post_search_products_response.json (hunk 9 above)
Comment (BirgitBader) [reply _raml/apps/search/examples/post_search_products_response.json]: 👍

File: apps/data-types.md (hunk 10 above)
Comment (BirgitBader) [reply apps/data-types.md]: 👍

File: apps/data-types.md (hunk 11 above)
Comment (BirgitBader) [reply apps/data-types.md]: 👍

File: apps/data-types.md (hunk 12 above)
Comment (BirgitBader) [reply apps/data-types.md]: well spotted

**Pull request merged**